extensions/amp-date-picker/1.0/component/date-range-picker.tsx

```diff
@@ -0,0 +1,432 @@
+import {
+  addDays,
+  differenceInDays,
+  isAfter,
+  isBefore,
+  isSameDay,
+  isValid,
+} from 'date-fns';
+import {DateRange, Matcher} from 'react-day-picker';
+
+import {Keys_Enum} from '#core/constants/key-codes';
+
+import {
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useMemo,
+  useRef,
+  useState,
+} from '#preact';
+import * as Preact from '#preact';
+import {forwardRef} from '#preact/compat';
+import {ContainWrapper} from '#preact/component';
+import {Ref} from '#preact/types';
+
+import {BaseDatePicker} from './base-date-picker';
+import {useDatePickerContext} from './use-date-picker-context';
+import {useDatePickerInput} from './use-date-picker-input';
+import {useDatePickerState} from './use-date-picker-state';
+
+import {
+  DateFieldType,
+  DateRangePickerAPI,
+  DateRangePickerProps,
+} from '../types';
+
+function DateRangePickerWithRef(
+  {
+    allowBlockedEndDate,
+    allowBlockedRanges,
+    children,
+    endInputSelector,
+    initialVisibleMonth,
+    mode,
+    openAfterClear,
+    openAfterSelect,
+    startInputSelector,
+    today,
+  }: DateRangePickerProps,
+  ref: Ref<DateRangePickerAPI>
+) {
+  const containerRef = useRef<HTMLElement>(null);
+
+  const [focusedInput, setFocusedInput] =
+    useState<DateFieldType>('start-input');
+
+  const defaultMonth = initialVisibleMonth || today;
+  const [month, setMonth] = useState<Date>(defaultMonth);
+
+  const {isOpen, transitionTo} = useDatePickerState(mode);
+
+  const {
+    blockedDates,
+    formatDate,
+    getDisabledAfter,
+    getDisabledBefore,
+    isDisabled,
+    parseDate,
+  } = useDatePickerContext();
+
+  const startDateInput = useDatePickerInput({
+    inputSelector: startInputSelector,
+    type: 'start-input',
+  });
+
+  const endDateInput = useDatePickerInput({
+    inputSelector: endInputSelector,
+    type: 'end-input',
+  });
+
+  /**
+   * Sets the selected date, month, and input value
+   */
+  const handleSetStartDate = useCallback(
+    (date?: Date) => {
+      if (!date) {
+        return;
+      }
+      startDateInput.handleSetDate(date);
+      setMonth(date);
+    },
+    [startDateInput]
+  );
+
+  /**
+   * Sets the selected date, month, and input value
+   */
+  const handleSetEndDate = useCallback(
+    (date?: Date) => {
+      if (!date) {
+        return;
+      }
+      endDateInput.handleSetDate(date);
+      setMonth(date);
+    },
+    [endDateInput]
+  );
```

Comment:
`handleSetStartDate` and `handleSetEndDate` seem extremely similar... I wonder if it'd be worthwhile to refactor them to something like 
```suggestion
  /**
   * Sets the selected date, month, and input value
   */
  const handleSetStartOrEndDate = useCallback(
    (date?: Date, startOrEndDate: 'start' | 'end') => {
      if (!date) {
        return;
      }
      if(startOrEndDate === 'start') {
         startDateInput.handleSetDate(date);
      } else {
        endDateInput.handleSetDate(date);
      }
      setMonth(date);
    },
    [startDateInput, endDateInput]
  );
```
We could probably make it even more simple then this, but this is just quick off the top of my head.

Reply:
They are pretty similar! In cases like this I prefer a little bit of repetition over introducing an extra conditional, but it's definitely a personal preference!

Reply:
I'm down for that.